token/js/test/unit/transferHook.test.ts

```diff
@@ -1,173 +1,547 @@
+            addressConfig,
+            isSigner: true,
+            isWritable: false,
+        };
+        const pdaExtraAccount = Buffer.concat([
+            Buffer.from([1]), // u8 discriminator
+            addressConfig, // 32 bytes address config
+            Buffer.from([1]), // bool isSigner
+            Buffer.from([0]), // bool isWritable
+        ]);
+
+        const pdaExtraAccountMetaWithProgramId = {
+            discriminator: 128,
+            addressConfig,
+            isSigner: false,
+            isWritable: true,
+        };
+        const pdaExtraAccountWithProgramId = Buffer.concat([
+            Buffer.from([128]), // u8 discriminator
+            addressConfig, // 32 bytes address config
+            Buffer.from([0]), // bool isSigner
+            Buffer.from([1]), // bool isWritable
+        ]);
+
+        const extraAccountList = Buffer.concat([
+            Buffer.from([0, 0, 0, 0, 0, 0, 0, 0]), // u64 accountDiscriminator
+            Buffer.from([0, 0, 0, 0]), // u32 length
+            Buffer.from([3, 0, 0, 0]), // u32 count
+            plainExtraAccount,
+            pdaExtraAccount,
+            pdaExtraAccountWithProgramId,
+        ]);
+
+        before(async () => {
+            connection = await getConnection();
+            connection.getAccountInfo = async (
+                _publicKey: PublicKey,
+                _commitmentOrConfig?: Parameters<(typeof connection)['getAccountInfo']>[1]
+            ): ReturnType<(typeof connection)['getAccountInfo']> => ({
+                data: Buffer.from([0, 0, 2, 2, 2, 2]),
+                owner: PublicKey.default,
+                executable: false,
+                lamports: 0,
+            });
+        });
+
+        it('can parse extra metas', () => {
+            const accountInfo = {
+                data: extraAccountList,
+                owner: PublicKey.default,
+                executable: false,
+                lamports: 0,
+            };
+            const parsedExtraAccounts = getExtraAccountMetas(accountInfo);
+            expect(parsedExtraAccounts).to.not.be.null;
+            if (parsedExtraAccounts == null) {
+                return;
+            }
+
+            expect(parsedExtraAccounts).to.have.length(3);
+            if (parsedExtraAccounts.length !== 3) {
+                return;
+            }
+
+            expect(parsedExtraAccounts[0].discriminator).to.eql(0);
+            expect(parsedExtraAccounts[0].addressConfig).to.eql(plainAccount.toBuffer());
+            expect(parsedExtraAccounts[0].isSigner).to.be.false;
+            expect(parsedExtraAccounts[0].isWritable).to.be.false;
+
+            expect(parsedExtraAccounts[1].discriminator).to.eql(1);
+            expect(parsedExtraAccounts[1].addressConfig).to.eql(addressConfig);
+            expect(parsedExtraAccounts[1].isSigner).to.be.true;
+            expect(parsedExtraAccounts[1].isWritable).to.be.false;
+
+            expect(parsedExtraAccounts[2].discriminator).to.eql(128);
```

Comment:
side note: I had totally forgotten about using the higher bits for the program id to use on the PDA, we'll definitely need that documented somewhere too

Reply:
Yeah, the constructors like `new_with_external_pda(..)` aren't in the JS lib since it's designed for reading extra meta configs. Creating them is pretty manual atm.

I'm assuming you mean document that in the Transfer Hook docs? Or do you mean in this library?

Reply:
Right, makes sense. and correct, I mean in the transfer hook docs, since JS lib users probably don't need to know about it